ledger-service/http-json/src/main/scala/com/digitalasset/http/PackageService.scala

```diff
@@ -44,36 +50,116 @@ private class PackageService(reloadPackageStoreIfChanged: PackageService.ReloadP
+          case None => logger.debug(s"new package IDs not found")
         }
-      }
-    }
+        .map { res =>
+          updateInstant(Instant.now())
+          res
+        }
+        .run
+  }
+
+  private val cache = StateCache()
+  private def state: State = cache.state
+
+  @inline
+  def reload(jwt: Jwt)(implicit
+      ec: ExecutionContext,
+      lc: LoggingContextOf[InstanceUUID],
+  ): Future[Error \/ Unit] = cache.reload(jwt)
 
   def packageStore: PackageStore = state.packageStore
 
   // Do not reduce it to something like `PackageService.resolveTemplateId(state.templateIdMap)`
   // `state.templateIdMap` will be cached in this case.
-  def resolveTemplateId: ResolveTemplateId =
-    x => PackageService.resolveTemplateId(state.templateIdMap)(x)
+  def resolveTemplateId(implicit ec: ExecutionContext): ResolveTemplateId = {
+    implicit lc: LoggingContextOf[InstanceUUID] => (jwt: Jwt) => (x: TemplateId.OptionalPkg) =>
+      {
+        type ResultType = Option[TemplateId.RequiredPkg]
+        def doSearch() = PackageService.resolveTemplateId(state.templateIdMap)(x)
+        def doReloadAndSearchAgain() = EitherT(reload(jwt)).map(_ => doSearch())
+        def keep(it: ResultType) = EitherT.pure(it): ET[ResultType]
+        for {
+          result <- EitherT.pure(doSearch()): ET[ResultType]
+          _ = logger.trace(s"Result: $result")
+          finalResult <-
+            x.packageId.fold {
+              if (result.isDefined)
+                // no package id and we do have the package, refresh if timeout
+                if (cache.packagesShouldBeFetchedAgain) {
+                  logger.trace(
+                    "no package id and we do have the package, refresh because of timeout"
+                  )
+                  doReloadAndSearchAgain()
+                } else {
+                  logger.trace("no package id and we do have the package, -no timeout- no refresh")
+                  keep(result)
+                }
+              // no package id and we don’t have the package, always refresh
+              else {
+                logger.trace("no package id and we don’t have the package, always refresh")
+                doReloadAndSearchAgain()
+              }
+            } { packageId =>
+              if (result.isDefined) {
+                logger.trace("package id defined & template id found, no refresh necessary")
+                keep(result)
+              } else {
+                // package id and we have the package, never refresh
+                if (state.packageIds.contains(packageId)) {
+                  logger.trace("package id and we have the package, never refresh")
+                  keep(result)
+                }
+                // package id and we don’t have the package, always refresh
+                else {
+                  logger.trace("package id and we don’t have the package, always refresh")
+                  doReloadAndSearchAgain()
+                }
+              }
```

Comment:
@cocreature can you take a look whether this is what you wanted and if not what should be changed?

Reply:
looks sensible :+1: